### ISSUE TITLE
Connecting from Node.js using 127.0.0.1:8335 or bchd.greyh.at:8335

### ISSUE THREAD

**Issue by arsen3d** [gcash/bchd#299]
If I attempt to use gRpc on bchd.greyh.at:8335 I have a problem with using 
grpc.client.submitTransaction()
I have the raw transaction which looks like this: 0100000001dabba7e2861bbbec9ece505d13074b47d062d75ffd4a297ddc977c2de5bfe9e40000000000ffffffff02e9030000000000001976a9140898a3ad5d87c63c82c6d00f2c8dc310f4713fce88ac04da1900000000001976a9148240a6d7bbc6f1ad88b59f1dde5d402661e9a4b788ac00000000

I constructed it like this:
```
import bitcore from "bitcore-lib-cash"; 

let t = bitcore.Transaction();
t.from(jsUtxos)
t.to("qqyf3gadtkruv0yzcmgq7tydcvg0guflec83sfdhdt", 1001)
t.change(myAddress)
t.sign(privateKey);
let rawTransaction = t.toString()
```
it would be nice if I could simply do something like:

```
import {GrpcClient} from "grpc-bchrpc-node";
let grpc = new GrpcClient(); 
grpc.client.submitTransaction(rawTransaction)
```

Instead I have to run my own Bchd server and submit the transaction via the JSON Rest api.
It looks like this:
```
let username = "myuser";
let password = "mypass";

var options ={
  headers:{
    "Authorization": `Basic ${(new Buffer(`${username}:${password}`).toString('base64'))}`,
    "Content-Type": "application/x-www-form-urlencoded"
  },
  data:JSON.parse(JSON.stringify( {"jsonrpc":"1.0","id":"curltext","method":"sendrawtransaction","params":[t.toString() ]}))
};

HTTP.call("POST",
  "http://127.0.0.1:8334",
  options,
  (e,r)=>{
    console.log(e,r)
  })
```
To do this I must set tls off in config file on my Bchd server notls=1

Setting notls=1 makes it so that I can only make requests from localhost.

Otherwise there will be this error:  { Error: write EPROTO 4500438336:error:1414D17A:SSL routines:tls12_check_peer_sigalg:wrong curve:../deps/openssl/openssl/ssl/t1_lib.c:1101:
when posting to:
https://127.0.0.1:8334
note that when notls=1 you must post to 
http://127.0.0.1:8334
instead

So I can’t use just remote server because it does not have JSONRpc enabled and gRPC does not have an interface for sending a raw transaction grpc.client.submitTransaction()

I am also unable to setup local instance of bchd to work without having to create and sign certificates.
I have not figured out how to authenticate to gRPC on localhost with notls=1 
It appears that notls=1 does not effect gRPC as upon attempting to connect I get the following error:
```
http: TLS handshake error from 127.0.0.1:52374: tls: first record does not look like a TLS handshake
```
I want to avoid all security certificates including letsEncrypt because I don’t want to worry about having to automating their creation and their renewal. I want communication via gRPC to be secured by the private network. Not that there is anything sensitive as transaction signing happens on the client.

In conclusion I must use two different bchd nodes to send a transaction. One to get UTXOs and the other to Broadcast. Also I intend to run my own bchd node in pruned mode with fastsync (prune=1, prunedepth=288, fastsync=1). This means that as of today UTXOs will not be available to me making it impossible to construct a transaction.

**Comment by cpacia** [comment]
> I have not figured out how to authenticate to gRPC on localhost with notls=1 

Well we can't use notls with grpc as grpc uses http2 which requires tls. 

Have you taken a look at the javascript examples? Afaik they work and should allow you to connect to the greyh.at server.

**Comment by arsen3d** [comment]
Which javascript examples?
I attempted to work with
https://github.com/simpleledgerinc/grpc-bchrpc-node
and submitted an issue expressing need to broadcast transactions using their lib
https://github.com/simpleledgerinc/grpc-bchrpc-node/issues/1

**Comment by arsen3d** [comment]
Here is my working example:
```
(async ()=>{
  let pks = process.env.pks;
  let pks2 =  process.env.pks2;
  let privateKey = new bitcore.PrivateKey(pks2);
  //console.log(privateKey.toJSON().bn)
  let myAddress = privateKey.toAddress().toString().split(":")[1];
  //let myAddress = "qpdjavghgulay9ym7rfd600dml5vlaqexguc482d62";//one utxo
  //let myAddress = "qrlqzlf7kghd2ju00tdm6yp7l5sz7ngkxccnaljsgq";// two utxos
  console.log("Looking up address: ", myAddress);
  console.log("Looking up legacy: ", privateKey.toAddress().toLegacyAddress());
  let utxosRaw ;

  //let grpc = new GrpcClient({url:"127.0.0.1:8335,"});//{url:"bchd.greyh.at:8335"}
  let grpc = new GrpcClient();

  utxosRaw = await grpc.getAddressUtxos("qpdjavghgulay9ym7rfd600dml5vlaqexguc482d62");
  console.log("https://blockchair.com/bitcoin-cash/address/qpdjavghgulay9ym7rfd600dml5vlaqexguc482d62 expected 1 UTXOs. Got: ", utxosRaw.toObject(false).outputsList.length)
  utxosRaw = await grpc.getAddressUtxos("qrlqzlf7kghd2ju00tdm6yp7l5sz7ngkxccnaljsgq");
  console.log("https://blockchair.com/bitcoin-cash/address/qrlqzlf7kghd2ju00tdm6yp7l5sz7ngkxccnaljsgq expected 2 UTXOs. Got: ", utxosRaw.toObject(false).outputsList.length)
  utxosRaw = await grpc.getAddressUtxos("pqeuuy6m8l5qkx9pqpevn2xrhqkfdl83y5yw6v9g97");
  console.log("https://blockchair.com/bitcoin-cash/address/pqeuuy6m8l5qkx9pqpevn2xrhqkfdl83y5yw6v9g97 expected 3 UTXOs. Got: ", utxosRaw.toObject(false).outputsList.length)
  // utxosRaw = await grpc.getAddressUtxos("pr7zc5jya6hsld6esavv6aplw6ttnt9hgg49229ld7");
  // console.log("https://blockchair.com/bitcoin-cash/address/pr7zc5jya6hsld6esavv6aplw6ttnt9hgg49229ld7 expected 407 UTXOs. Got: ", utxosRaw.toObject(false).outputsList.length)

  //myAddress = "qrlqzlf7kghd2ju00tdm6yp7l5sz7ngkxccnaljsgq"
  utxosRaw = await grpc.getAddressUtxos(myAddress);
  console.log("https://blockchair.com/bitcoin-cash/address/qqyf3gadtkruv0yzcmgq7tydcvg0guflec83sfdhdt expected 1 UTXOs. Got: ", utxosRaw.toObject(false).outputsList.length)


  let jsUtxos = [];
  for (let x  = 0;x<utxosRaw.toObject(false).outputsList.length; x++){
    let id = Buffer.from(utxosRaw.getOutputsList()[x].getOutpoint().getHash_asU8().reverse()).toString("hex");
    let script =  Buffer.from(utxosRaw.getOutputsList()[x].getPubkeyScript_asU8()).toString("hex");
    let value = utxosRaw.toObject(false).outputsList[x].value;
    console.log("My Address: ",myAddress );
    console.log("Output TX: ",id );
    console.log("Script: ",script);
    console.log("Value: ",value);
    let utxo = {}
    utxo.txId = id;
    utxo.outputIndex= x;
    utxo.address= myAddress;
    utxo.script=  script;
    utxo.satoshis= value;
    jsUtxos.push(utxo)
  }
if(jsUtxos.length == 0)
{
  console.log("No UTXOS Check address")
  return;
}


console.log("UTXOs: ",JSON.stringify(jsUtxos,null,2));
  let t = bitcore.Transaction();
  t.from(jsUtxos)
  t.to("qqyf3gadtkruv0yzcmgq7tydcvg0guflec83sfdhdt", 1001)
  t.change(myAddress)
  t.sign(privateKey);

  console.log("Is verify: ",t.verify())
  console.log("isFullySigned: ",t.isFullySigned())


  console.log("RAW Transaction to broadcast ",t.toString());
  //console.log("utxoDecoder.decode:", JSON.stringify( utxoDecoder.decode(t.toString()),null,2));

  let username = "myuser";
  let password = "mypass";

  var options ={
    headers:{
      "Authorization": `Basic ${(new Buffer(`${username}:${password}`).toString('base64'))}`,
      "Content-Type": "application/x-www-form-urlencoded"
    },
    data:JSON.parse(JSON.stringify( {"jsonrpc":"1.0","id":"curltext","method":"sendrawtransaction","params":[t.toString() ]}))
  };
  HTTP.call("POST",
    "https://127.0.0.1:8334",
    options,
    (e,r)=>{
      console.log(e,r)
    })

})()
```

**Comment by arsen3d** [comment]
I managed to connect with gRPC to localhost host by using this file, which was autogenerated
/Users/arsen/Library/Application Support/Bchd/rpc.cert
Got the following error:
Error: 14 UNAVAILABLE: addrindex required
When calling:
```
 process.env.NODE_TLS_REJECT_UNAUTHORIZED = 0;
  let grpc = new GrpcClient({url:"127.0.0.1:8335",rootCertPath:"/Users/arsen/Library/Application Support/Bchd/rpc.cert"});
  let utxosRaw = await grpc.getAddressUtxos("qpdjavghgulay9ym7rfd600dml5vlaqexguc482d62");
  assert.strictEqual(utxosRaw.toObject(false).outputsList.length,1)
```
So I added 
addrindex=1
to bchd.conf

Then I got:
loadConfig: txindex and addrindex can not be used with fast sync mode.

so I set fastsync=0

Then I got:
19-11-05 20:20:19.404 [INF] INDX: Transaction index enabled because it is required by the address index
2019-11-05 20:20:19.404 [INF] INDX: Address index is enabled
2019-11-05 20:20:19.404 [INF] INDX: Committed filter index is enabled
2019-11-05 20:20:19.405 [INF] CHAN: Loading block index...
2019-11-05 20:20:22.868 [INF] INDX: Catching up indexes from height -1 to 607787
2019-11-05 20:20:23.045 [ERR] BCHD: Unable to start server on [:8333]: assertion failed: mismatched spend journal serialization - no serialization for expected 1 stxos


My goal is to run config with  prune=1, prunedepth=288 , fastsync=1, addrindex=1

Basically a fully functional node without the huge SSD cloud storage requirement

**Comment by cpacia** [comment]
https://github.com/gcash/bchd/tree/master/bchrpc/documentation/client-usage-examples/nodejs-grpc

**Comment by arsen3d** [comment]
> https://github.com/gcash/bchd/tree/master/bchrpc/documentation/client-usage-examples/nodejs-grpc

I tried to use 
process.env.NODE_TLS_REJECT_UNAUTHORIZED = 0;
var client = new pb.bchrpc('127.0.0.1:8335', grpc.credentials.createSsl());
without passing in a certificate the authentication still failed 
with http: TLS handshake error from 127.0.0.1:55815: EOF on bchd 
and 
(STDERR) E1105 20:45:08.267764000 4495809856 ssl_transport_security.cc:1237]    Handshake failed with fatal error SSL_ERROR_SSL: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed.
on node.
Here is the code:
```
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = 0;
  let grpc = new GrpcClient({url:"127.0.0.1:8335"}); //,rootCertPath:"/Users/arsen/Library/Application Support/Bchd/rpc.cert"
  let utxosRaw = await grpc.getAddressUtxos("qpdjavghgulay9ym7rfd600dml5vlaqexguc482d62");
  assert.strictEqual(utxosRaw.toObject(false).outputsList.length,1)

```
and modified version of "grpc-bchrpc-node"
```
var creds = grpc.credentials.createSsl();
        console.log(creds)
        if (rootCertPath) {
            var rootCert = fs.readFileSync(rootCertPath);
            console.log(rootCert)
            creds = grpc.credentials.createSsl(rootCert);
        }
        if (!url && !testnet) {
            creds = grpc.credentials.createSsl();
            url = "bchd.greyh.at:8335";
        }
        else if (!url) {
            url = "bchd-testnet.greyh.at:18335";
        }
        if (!options) {
            options = {
                'grpc.max_receive_message_length': -1,
            };
        }
        console.log(url)
        this.client = new bchrpc_grpc.bchrpcClient(url, creds, options);
```

**Comment by arsen3d** [comment]
Some notes on the topic:
```
zquestz Today at 10:33 AM
The UTXO data needs a fully indexed node, it requires an addrindex and we are working on a lightweight utxo address index, but that hasn't been done yet. (edited) 

5 replies

Arsen Yeremin  10 hours ago
How will you build the index without downloading all the transactions first?

tyler  10 hours ago
the addrindex and txindex require downloading all txs. the utxoaddr index will be the only one that works with a fastsync node because it can index from the UTXO set directly.

Arsen Yeremin  7 hours ago
So is fully functional pruned node expected on the roadmap this year?
Also sent to the channel

tyler  7 hours ago
I’m not sure what you mean exactly. Pruning on your own node will still download all txs, it just won’t store them forever. So your node can still build all the indexes.
If you fastsync then you skip downloading all the txs and there’s no way possible to generate the indexes. It’s a trade off you make to have fast sync times. It’s possible we could build a system to distributed serialized indexes but they’d have to be trusted since there’s no way to validate them without rebuilding them from scratch.

Arsen Yeremin  7 hours ago
Understood.  Deploying a node to the cloud requires provisioning a server with a set size SSD. If the BCH is a few tarabytes in a few years it would be good not to have to provision that much  space to initially download the whole thing. If the whole thing has to be rebuilt from scratch, it would be good to purge the validated transactions as they are checked.
```

**Comment by jcramer** [comment]
@arsen3d please see issue #330, I was able to get it working with a self-signed cert.  But I agree it would be really nice if the auto-generated `rpc.cert` file worked out of the gate.  Anywho, the workaround isn't that bad.

**Comment by zquestz** [comment]
@arsen3d I will be committing a fix for the issue @jcramer describes. That should help a lot with connecting from nodejs!

**Comment by zquestz** [comment]
This is resolved. Thanks for the report.

**Comment by FreeTrade** [comment]
@zquestz 

Think i'm still running into this error - using the certs created by 0.16.5 -

on the client end i get 
Error: 14 UNAVAILABLE: failed to connect to all addresses

on the bchd end i get
http: TLS handshake error from 127.0.0.1:64007: EOF

seems similar to @arsen3d 's problem above
